[Yarr] Improve processing of adjacent or near adjacent single characters

msaboff opened the pull request:
#### 1e14cbbdc2f52569760d7b86f3b0cec9709207ce
<pre>
[Yarr] Improve processing of adjacent or near adjacent single characters
<a href="https://bugs.webkit.org/show_bug.cgi?id=289567">https://bugs.webkit.org/show_bug.cgi?id=289567</a>
<a href="https://rdar.apple.com/problem/146795365">rdar://problem/146795365</a>

Reviewed by Yusuke Suzuki.

Updated the multi-character load, compare and branch code in generatePatternCharacterOnce() to
consider all single character and single width character classes as a group.  Given that these terms
may appear out of order due to other optimizations, we put the terms in order of their character
position in the alternation.  We then take the leading contiguous terms together, stopping at the
first non single width character or character class term or if there is a missing term for a
character position (like for a non-BMP codepoint).

We then create the largest width load, compare and branch sequence which may include masking out
the positions for character class terms that will be handled separately.  In some cases, we perform
overlapping loads.  For example if we have 7 characters to process, we can perform two 4 byte load
and compares that overlap with the middle character.

Overall, this reduces the number of load, compare and branch sequences.  This improvement significantly
helps the processing of RegExps with strings or strings with embedded single width character classes.
The SunSpider regex-dna subtest is a good example of the benefits of this improvement:

                         Baseline                 GroupChars

regexp-dna            4.1290+-0.1258     ^      2.8660+-0.1270        ^ definitely 1.4407x faster

&lt;arithmetic&gt;          4.1290+-0.1258     ^      2.8660+-0.1270        ^ definitely 1.4407x faster

Added tests that exercise the new code.

* JSTests/stress/regexp-multiple-char-matching.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
(printErrors):
(let.re.break.case.catch.continue.debugger.default.else.finally.if):
(RegExpTest):
(RegExpTest.prototype.runTest):
(RegExpTestList):
(RegExpTestList.prototype.addTest):
(RegExpTestList.prototype.runTests):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:

Canonical link: <a href="https://commits.webkit.org/292003@main">https://commits.webkit.org/292003@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e035bee5c55c6ffee4e1983f0657f0ec1c75d7b6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94647 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14238 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4090 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99668 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45146 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14519 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72196 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29499 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97649 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10807 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85470 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52530 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10500 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3180 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44482 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/87317 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80718 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3280 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101710 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/93275 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21683 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15818 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81193 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21931 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81501 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80574 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20116 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25133 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14900 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21662 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26782 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115961 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21332 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24797 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23070 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -1993,252 +1994,373 @@ class YarrGenerator final : public YarrJITInfo {
-        if (op.m_term->ignoreCase() && isASCIIAlpha(ch)) {
+        if (have16BitCharacter && m_charSize == CharSize::Char16 && !U_IS_BMP(firstChar)) {
+            // The first term we are considering is a non-BMP char in a 16 bit pattern. Just try matching it and be done.
 #if CPU(BIG_ENDIAN)
```
We no longer support BigEndian JIT architecture, so maybe we can just remove this code.

msaboff replied:
I will gladly remove the BIG_ENDIAN code.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -1993,252 +1994,373 @@ class YarrGenerator final : public YarrJITInfo {
+            if (currTerm->quantityType != QuantifierType::FixedCount
+                || currTerm->quantityMaxCount != 1
+                || (currTerm->type != PatternTerm::Type::PatternCharacter
+                    && currTerm->type != PatternTerm::Type::CharacterClass)
```
OK, now it is extended for character class.

msaboff replied:
We treat these character classes as holes that we mask around when processing the surrounding pattern characters.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -1993,252 +1994,373 @@ class YarrGenerator final : public YarrJITInfo {
+                    else
+                        matchTargets.appendSucceeded(m_jit.branch32(MacroAssembler::Equal, character, MacroAssembler::Imm32(characters | caseMask)));
+                };
 #endif
```
Removing dead code?

msaboff replied:
I removed this in the reposted PR.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -1993,252 +1994,361 @@ class YarrGenerator final : public YarrJITInfo {
 #if CPU(BIG_ENDIAN)
-            int shiftAmount = (m_charSize == CharSize::Char8 ? 24 : 16) - ((m_charSize == CharSize::Char8 ? 8 : 16) * numberCharacters);
+                return (m_charSize == CharSize::Char8 ? 24 : 16) - ((m_charSize == CharSize::Char8 ? 8 : 16) * positionInLoad);
 #else
```
Ditto.

msaboff replied:
Done

msaboff committed:
[Yarr] Improve processing of adjacent or near adjacent single characters
https://bugs.webkit.org/show_bug.cgi?id=289567
rdar://problem/146795365

Reviewed by Yusuke Suzuki.

Updated the multi-character load, compare and branch code in generatePatternCharacterOnce() to
consider all single character and single width character classes as a group.  Given that these terms
may appear out of order due to other optimizations, we put the terms in order of their character
position in the alternation.  We then take the leading contiguous terms together, stopping at the
first non single width character or character class term or if there is a missing term for a
character position (like for a non-BMP codepoint).

We then create the largest width load, compare and branch sequence which may include masking out
the positions for character class terms that will be handled separately.  In some cases, we perform
overlapping loads.  For example if we have 7 characters to process, we can perform two 4 byte load
and compares that overlap with the middle character.

Overall, this reduces the number of load, compare and branch sequences.  This improvement significantly
helps the processing of RegExps with strings or strings with embedded single width character classes.
The SunSpider regex-dna subtest is a good example of the benefits of this improvement:

                         Baseline                 GroupChars

regexp-dna            4.1290+-0.1258     ^      2.8660+-0.1270        ^ definitely 1.4407x faster

<arithmetic>          4.1290+-0.1258     ^      2.8660+-0.1270        ^ definitely 1.4407x faster

Added tests that exercise the new code.

* JSTests/stress/regexp-multiple-char-matching.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
(printErrors):
(let.re.break.case.catch.continue.debugger.default.else.finally.if):
(RegExpTest):
(RegExpTest.prototype.runTest):
(RegExpTestList):
(RegExpTestList.prototype.addTest):
(RegExpTestList.prototype.runTests):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:

Canonical link: https://commits.webkit.org/292003@main
